### PR TITLE
feat(vocab): skip Geneve option section to resolve inner offset

### PR DESCRIPTION
**Pull request by takehaya**
## 概要

このPRは当初「Geneve option section を skip して inner offset を解決する」変更として始まったが、実装の過程で**より深い offset モデルの correctness バグ**が判明し、その修正を含む形に拡張された。

### 背景バグ（核心）
`markRuntimeOffsetLayers` は heterogeneous-size alternation しか runtime-offset 境界として扱っておらず、**可変長レイヤ（ipv4 options / gtp opt・exts / ipv6 exts / srv6 segments / geneve options）より後ろのレイヤへの `where`/capture 読みが固定 offset に解決**され、可変部が非空のとき誤読していた。Geneve の `where inner.dst` がこれを初めて踏んだ。

### 変更内容（commit 単位）
1. `feat(vocab): skip Geneve option section` — geneve.p4 に `skip_options` state（`opt_len*4` byte を field-driven advance、SRv6 と同じ idiom）。
2. `test(dsltest): validate Geneve OptLenWords range` — テストビルダの range 検証。
3. `fix(codegen): runtime offset for layers past a variable-length predecessor` — **本体修正**。可変長レイヤも境界として扱い、downstream の where/capture 読みを per-layer slot 経由に。dynamic aux-index 経路を `layerAnchorFor` に移行、dead code 削除。
4. `refactor(resolve): name the runtime-offset boundary predicate` — `isRuntimeBoundary` 抽出（純粋リファクタ）。
5. `fix(resolve): treat a variable-layout alt member as a runtime boundary` — 固定長は揃うがメンバーが可変長の alt（例 `(vxlan|geneve)`）も境界に。回帰テスト追加。
6. `test(resolve): isolate the variable-layout alt-member boundary case` — 上記テストを udp-root にして分離。

### 命令数（raw insns）の変化
offset 修正で downstream 読みを持つ filter は per-layer slot store の分だけ増える。**最終的な pin 値**:

| | 旧 | 新 |
|---|---:|---:|
| F1 | 197 | **199** |
| F2 | 207 | **209** |
| F3 | 338 | 338 (不変) |
| F4 | 213 | **215** |
| F5 | 215 | **217** |
| F6 | 83 | **85** |
| F7 | 403 | **405** |
| F8 | 425 | **450** |
| F9 | 299 | **323** |
| F10 | 230 | **231** |

> F9 の内訳: geneve option skip で 299→321（+22）、offset モデル修正の slot store で 321→**323**（+2）。中間 commit のメッセージにある「299→321」はこの段階値で、**最終 stable は 323**。

### 検証
- CI: BPF load test x5 kernel (6.1/6.6/6.12/6.18/7.0) + Unit + Integration + lint + p4c **全 pass**。
- ローカル (kernel 6.15): 全 dsltest packet-level + filterset/corpus verifier load pass。
- この offset 修正は副次的に F7（GTP-U inner, extension 有り）等の可変長チェインの latent 誤読も是正する。

### 論文・bench への波及（このPR外、別途対応）
- `benchmark/results/b1_insns.csv` + Fig.4 の dsl 命令数行は再生成が必要（cbpfc 列は不変）。
- 論文の命令数テーブルは更新済み（paper ブランチ、未コミット）。cbpfc 比 2.7-14.7x は不変を確認済み。

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/kunai/dsltest/builders.go [hunk 1]
```diff
@@ -597,6 +603,11 @@ func BuildGeneveInnerIPv4TCP(t testing.TB, opts GeneveInnerIPv4TCPOpts) []byte {
 	_ = udp.SetNetworkLayerForChecksum(v4)
 
 	geneve := geneveHeader(opts.VNI, layers.EthernetTypeTransparentEthernetBridging)
+	if opts.OptLenWords > 0 {
+		// version stays 0; opt_len occupies the low 6 bits of byte 0.
+		geneve[0] = byte(opts.OptLenWords & 0x3F)
+		geneve = append(geneve, make([]byte, opts.OptLenWords*4)...)
+	}
```
[the remark this answers is not in the document]
Comment (takehaya) [reply pkg/kunai/dsltest/builders.go]: Fixed in db1b260. Added a fail-fast range check (OptLenWords is a 6-bit field, 0..63) and now write the single validated value into both the opt_len byte and the filler length, dropping the partial `& 0x3F` mask so header and payload can't desync.

File: pkg/kunai/resolve/resolve.go [hunk 2]
```diff
@@ -170,7 +176,13 @@ func addChainRootWarning(p *ir.Program, f *ast.Filter) {
 // into the same slot — codegen's per-alt advance logic guarantees
 // whichever alt matched is the one whose R4 entry was just stored.
 func markRuntimeOffsetLayers(p *ir.Program) {
-	hetAltPos := -1
+	// A layer makes every later layer's start offset runtime-variable
+	// when it is a heterogeneous-size alternation or has a variable-
+	// length body (parser-machine trailer / pkt.advance skip).
+	isRuntimeBoundary := func(l *ir.LayerInstance) bool {
+		return ir.IsHeterogeneousAlt(l) || (l.Spec != nil && l.Spec.HasVariableLayout())
+	}
```
[the remark this answers is not in the document]
Comment (takehaya) [reply pkg/kunai/resolve/resolve.go:193]: Good catch — fixed in eaa89b1. markRuntimeOffsetLayers now also checks the alt members (slices.ContainsFunc over hasVarBody), so a uniform-size alt with a variable-layout member like `(vxlan|geneve)` is treated as a runtime boundary and downstream where/capture reads resolve through the per-layer slot. Added TestResolveMarksRuntimeOffsetForVariableLayoutAltMember.

File: pkg/kunai/resolve/resolve_test.go [hunk 3]
```diff
@@ -429,22 +429,50 @@ func TestResolveMarksRuntimeOffsetForLayersPastHetAlt(t *testing.T) {
...
-	// uniform-size (4 bytes each), so the static prefix path still
-	// works; no layer should be marked.
+func TestResolveMarksRuntimeOffsetForVariableLayoutAltMember(t *testing.T) {
+	// `eth/ipv4/udp/(vxlan|geneve)/eth/ipv4@inner/tcp where inner.dst`
+	// — vxlan and geneve share the 8-byte fixed size, so the alt is NOT
+	// heterogeneous, but geneve carries an opt_len-driven options
+	// trailer. The alt is therefore a runtime-offset boundary on the
+	// geneve branch, so the inner ipv4 read by the where clause must be
+	// marked NeedsRuntimeOffset.
+	p := resolveOK(t, "eth/ipv4/udp/(vxlan|geneve)/eth/ipv4@inner/tcp where inner.dst == 10.0.0.1", nil)
```
[the remark this answers is not in the document]
Comment (takehaya) [reply pkg/kunai/resolve/resolve_test.go]: Fixed in d6ddb15. Re-rooted the test at `udp/(vxlan|geneve)/eth/ipv4@inner/tcp` so every pre-alt layer is fixed-size and the alt is the first runtime boundary — now @inner is marked only via the alt-member check, so a regression in slices.ContainsFunc(l.Alternation, hasVarBody) makes the test fail. Verified: drop the alt-member check and @inner is no longer marked.

File: internal/program/filterset_test.go [hunk 4]
```diff
@@ -44,31 +44,31 @@ type FilterSpec struct {
...
+		WantInsns: 405, Notes: "GTP-U inner IPv4 dst (5G core)"},
 	{ID: "F8", Expr: "eth/ipv6/srv6 where any(srv6.segments.addr == fc00::1)",
-		WantInsns: 425, Notes: "SRv6 segment-list any-quantifier (5G/SDN)"},
+		WantInsns: 450, Notes: "SRv6 segment-list any-quantifier (5G/SDN)"},
 	{ID: "F9", Expr: "eth/ipv4@outer/udp/geneve/eth/ipv4@inner/tcp where inner.dst == 10.0.0.1",
-		WantInsns: 299, Notes: "Geneve inner IPv4 dst (Cilium overlay / cloud VPC)"},
+		WantInsns: 323, Notes: "Geneve inner IPv4 dst (Cilium overlay / cloud VPC); inner offset resolved past the opt_len*4 options section"},
 	{ID: "F10", Expr: "eth/ipv4/tcp where tcp.options.MSS.value == 1460",
-		WantInsns: 230, Notes: "TCP option MSS via parser-counter walk"},
+		WantInsns: 231, Notes: "TCP option MSS via parser-counter walk"},
```
[the remark this answers is not in the document]
Comment (takehaya) [reply internal/program/filterset_test.go:71]: 323 is the correct stable count. The 299→321 figure was the geneve-skip commit alone (+22 insns); the later offset-model commit adds the per-layer runtime-offset slot store (+2), giving 323. CI pins/loads it green across all 5 kernels. I'll reconcile the PR description; the pinned value 323 is intentional.

File: internal/program/filterset_test.go [hunk 5]
```diff
@@ -44,31 +44,31 @@ type FilterSpec struct {
...
+		WantInsns: 85, Notes: "ICMP echo request"},
 	{ID: "F7", Expr: "eth/ipv4@outer/udp/gtp/ipv4@inner/tcp where inner.dst == 10.0.0.1",
-		WantInsns: 403, Notes: "GTP-U inner IPv4 dst (5G core)"},
+		WantInsns: 405, Notes: "GTP-U inner IPv4 dst (5G core)"},
 	{ID: "F8", Expr: "eth/ipv6/srv6 where any(srv6.segments.addr == fc00::1)",
-		WantInsns: 425, Notes: "SRv6 segment-list any-quantifier (5G/SDN)"},
+		WantInsns: 450, Notes: "SRv6 segment-list any-quantifier (5G/SDN)"},
 	{ID: "F9", Expr: "eth/ipv4@outer/udp/geneve/eth/ipv4@inner/tcp where inner.dst == 10.0.0.1",
-		WantInsns: 299, Notes: "Geneve inner IPv4 dst (Cilium overlay / cloud VPC)"},
+		WantInsns: 323, Notes: "Geneve inner IPv4 dst (Cilium overlay / cloud VPC); inner offset resolved past the opt_len*4 options section"},
```
[the remark this answers is not in the document]
Comment (takehaya) [reply internal/program/filterset_test.go:69]: Updated the PR description: the table now lists 323 as F9's final stable count, with the 299→321→323 breakdown (geneve skip +22, then the offset-model slot store +2). The pinned 323 matches the code and is green across all 5 kernels.
